File: src/nuitrack_bridge/types/user_frame.cc
#include "nuitrack_bridge/types/user_frame.h"
#include "nuitrack/types/UserFrame.h"

namespace nuitrack_bridge::user_frame {

    std::unique_ptr<std::vector<User>> getUsers(const UserFrame& frame) {
        // Get the original vector from the Nuitrack SDK
        const auto sdk_users = frame.getUsers();
        
        // Create a new vector to hold the bridge-compatible User structs
        auto bridge_users = std::make_unique<std::vector<User>>();
        bridge_users->reserve(sdk_users.size());

        // Since the Rust-defined `User` has an identical memory layout to the SDK's `tdv::nuitrack::User`,
        // we can safely cast and copy each element.
        for (const auto& sdk_user : sdk_users) {
            bridge_users->push_back(*reinterpret_cast<const User*>(&sdk_user));
        }
        
        return bridge_users;
    }

    int32_t getRows(const UserFrame& frame) {
        return static_cast<int32_t>(frame.getRows());
    }

    int32_t getCols(const UserFrame& frame) {
        return static_cast<int32_t>(frame.getCols());
    }

    uint64_t getID(const UserFrame& frame) {
        return frame.getID();
    }

    rust::Slice<const uint16_t> getData(const UserFrame& frame) {
        return rust::Slice<const uint16_t>{
            frame.getData(),
            static_cast<size_t>(frame.getRows() * frame.getCols())
        };
    }

    uint64_t getTimestamp(const UserFrame& frame) {
        return frame.getTimestamp();
    }

    Vector3 getFloor(const UserFrame& frame) {
        auto sdk_vector = frame.getFloor();
        // Cast the SDK's Vector3 to the bridge's layout-compatible Vector3
        return *reinterpret_cast<Vector3*>(&sdk_vector);
    }

    Vector3 getFloorNormal(const UserFrame& frame) {
        auto sdk_vector = frame.getFloorNormal();
        // Cast the SDK's Vector3 to the bridge's layout-compatible Vector3
        return *reinterpret_cast<Vector3*>(&sdk_vector);
    }

} // namespace nuitrack_bridge::user_frame

File: src/nuitrack_bridge/types/user_frame.rs
#[cxx::bridge(namespace = "nuitrack_bridge::user_frame")]
pub mod ffi {
    // --- Rust <-> C++ Type Declarations ---

    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        /// An opaque type representing the Nuitrack UserFrame.
        type UserFrame;
    }

    #[namespace = "nuitrack_bridge::user"]
    unsafe extern "C++" {
        /// The User struct defined in user.rs
        type User = crate::nuitrack_bridge::types::user::ffi::User;
    }

    #[namespace = "nuitrack_bridge::vector3"]
    unsafe extern "C++" {
        /// The Vector3 struct defined in vector3.rs
        type Vector3 = crate::nuitrack_bridge::types::vector3::ffi::Vector3;
    }

    // Enable passing UserFrame in a smart pointer
    impl SharedPtr<UserFrame> {}
    // Enable receiving a vector of User structs
    impl CxxVector<User> {}

    // --- Function Signatures ---

    unsafe extern "C++" {
        include!("nuitrack_bridge/types/user_frame.h");

        #[cxx_name = "getUsers"]
        pub fn users(frame: &UserFrame) -> Result<UniquePtr<CxxVector<User>>>;

        #[cxx_name = "getRows"]
        pub fn rows(frame: &UserFrame) -> Result<i32>;

        #[cxx_name = "getCols"]
        pub fn cols(frame: &UserFrame) -> Result<i32>;
        
        #[cxx_name = "getID"]
        pub fn id(frame: &UserFrame) -> Result<u64>;

        #[cxx_name = "getData"]
        pub fn data<'a>(frame: &'a UserFrame) -> Result<&'a [u16]>;

        #[cxx_name = "getTimestamp"]
        pub fn timestamp(frame: &UserFrame) -> Result<u64>;
        
        #[cxx_name = "getFloor"]
        pub fn floor(frame: &UserFrame) -> Result<Vector3>;

        #[cxx_name = "getFloorNormal"]
        pub fn floor_normal(frame: &UserFrame) -> Result<Vector3>;
    }
}

// Mark the opaque C++ type as safe to send across threads
unsafe impl Send for ffi::UserFrame {}
unsafe impl Sync for ffi::UserFrame {}